### DISCUSSION TITLE
RFC for Magic sign-in link via email

### DISCUSSION THREAD

**Discussion by adanielyan** [directus/directus#17888] [Feature Requests]
### Summary

Add a "Magic sign-in link" option to the available authentication methods. Users must be able to choose a Magic sign-in option in addition to any other enabled option to sign in. This authentication method should work for both the Directus app and authentication through API. This authentication should be an extension of default (database) authentication. That is, the auth method on the account will be "default". As a result only accounts with a "default" authentication method will be able to use Magic link auth, although it may change in the future when/if authentication by any method (regardless of what method was used to create an account) is supported by Directus.

### Example 1
1. User opens the Directus app, enters email and clicks on "Send me a magic link" link/button
2. Directus generates a magic link and emails it to the user.
3. User clicks on the link in the email that authenticates them and redirects back to Directus.

### Example 2
1. User opens a custom app that has uses Directus user authentication under the hood.
2. User enters email and clicks on "Send me a magic link" link/button.
3. Directus generates a magic link and emails it to the user.
4. User clicks on the link in the email that authenticates them and redirects back to Directus.

### Motivation

Magic link authentication is becoming increasingly popular and are considered a safe way to authenticate user without password. It would be good to support this authentication method.

### Use case 1
1. User opens the directus app
2. If Magic link auth method is enabled, user sees a "Send me a magic link" link/button under the Email field
<img width="454" alt="image" src="https://user-images.githubusercontent.com/861777/226976159-42c0446b-e2ec-4e54-a70e-995f65a735e5.png">

3. User enters email and clicks on `  Send me a magic link  ` button. 
4. If a user with a given email exists, Directus automatically authenticates that user and generates a refresh token.
5. If user doesn't exist, Directus creates one and generates a refresh token, similar to OAuth authentication.
6. A URL containing a Directus token refresh API endpoint URL, the redirect URL (that is the Directus instance URL), and a refresh token is generated. 
7. That URL is then emailed to the email address provided by the user.
8. A message appears confirming that the email was sent.
9. User clicks on the link in the email that takes them to the token refresh API endpoint (see the drawbacks section below). 
10. If token is valid the token refresh API does its job and redirects user back to a redirect URL with a cookie that has a new access and refresh tokens (similar to regular API token refresh).

### Use case 2
1. User opens a custom app that has uses Directus user authentication.
2. If Magic link auth method is enabled and added to the app's login form, user sees an email field and "Send me a magic link" link/button. The UX implementation will be up to the app's owner
3. A request with user email and the current page URL is sent to a new Directus API endpoint.
Steps 4-10 are same as in `Use case 1`, but the redirect URL in step 6 will be the app's page URL submitted to the Magic Link generation API endpoint.

### Must Have:
- Token refresh API endpoint that accepts GET requests

### Should Have:
- To support magic link generation through API a Magic link generation API endpoint should be created

### Could Have:
- ?

### Won't Have:
- ?

### Drawbacks

To implement this approach the token refresh API endpoint will have to accept GET requests, as you cannot send a POST request from an email. While this may sound as a security risk, in reality it is not, as the refresh tokens can only be used once, and are discarded after a new refresh token is generated. But there may be other negative consequences that I cannot think of at the moment. 

### Alternatives

Using 3rd party OAuth providers, such as Auth0 to enable Magic link authentication.

### Adoption Strategy

TBD

**Comment by akiarostami** [comment]
For passwordless entry, I would also love to have the OTP option, so instead of a link, the user would get a 6 (or more) digit number to log in.

**Reply by rijkvanzanten** [reply]
You mean that the OTP is sent as an email still, or do you mean like two-factor authentication where the user enters their login information (email/password) and then an OTP for added security?

**Reply by akiarostami** [reply]
I mean OTP sent to an email for passwordless entry, and not TFA. Just like a magic link (instead of a link) a number is emailed to the user. With that, there would be no need to sign in via URI, and there would be no redirecting the user after auth. User stays in the log-in form with a field to enter the OTP number and login. Later on this can be extended to receiving a text message with that 6-digit number instead of email.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by w0kyj** [reply]
Related Feature Requests:
- https://github.com/directus/directus/discussions/18755
- https://github.com/directus/directus/discussions/13163

**Reply by BorisKamp** [reply]
What is the current status of this @rijkvanzanten ?

**Comment by rijkvanzanten** [comment]
Thanks @adanielyan! The front-end steps make a ton of sense. I think the last piece of the puzzle for this feature to be a success is figuring out the exact API spec. Couple of questions on that:

- How do you request this link to be sent?
- How do we control where the login link redirects you to (so you can magic-link back to your own app)
  - How do you configure what URLs are allowed (probably same as SSO?)
- How do we unset/expire the token after the login has happened?
  - Quick thought: potentially using a temporary session in directus_sessions just for the login

**Comment by vanling** [comment]
> Simplified From frontend app/development POV using SSO: 
> 
> On frontend -> trigger "email login"
> -> trigger something  as -> `GET /auth/login/:provider` with redirect param (that would be whitelisted (like forgot password))
> 
> -> Directus sends the email + link (Directus saves the redirect state on the user on the temp token?)
> -> user clicks the link in the email
> -> link goes to Directus and login request gets handled and redirects back to app
> 
> ☝️ this + a poopyload of more details 😛

**Comment by w0kyj** [comment]
Must have:
- support for custom apps (allow custom implementations to use as auth)

**Comment by ermincelikovic** [comment]
Hi everyone - I just wanted to check if there is any progress on this?

**Comment by jackmurraybrown** [comment]
Any update on the progress of this?

**Comment by lucasiscovici** [comment]
Hello, what is the status please ? @rijkvanzanten

**Comment by technoknol** [comment]
was there any progress on this ?

**Comment by BorisKamp** [comment]
Bumping this as well!

**Comment by carwei** [comment]
Me too. Or perhaps a tutorial on how to set this up until it's officially supported?

EDIT: I think I have solved this now

This points as a possibility https://docs.directus.io/blog/passwordless-sms-authentication-with-plivo-and-directus-automate.html but it would be nice to do it without 3rd parties (Plivo is not supported in all countries, and it's an extra GDPR "dependency").

**Reply by RawmanGrey** [reply]
Getting passwordless authentication with Directus was not a breeze. I've not attempted to implement it in Directus UI, but I had to authenticate users on bare phone number on separate resource using Directus SDK. 

To get things working I've added 'phone' and 'secret' field to directus_users. 
When user provides phone, server searches for Directus user by phone.
Then server generates OTP with [otplib](https://www.npmjs.com/package/otplib) and sends temporary access code via[ android-sms-gateway](https://docs.sms-gate.app/). Code could be delivered via email or any other SMS-gateway service.
If user provides correct OTP code to my service,  I have to authenticate him server-side in Directus strictly by email and password user doesn't have to know.

**Reply by carwei** [reply]
Good to know. For my use case I need magic links (user needs the low friction of clicking a link and might not have a phone).

**Comment by carwei** [comment]
I get that everyone is waiting for an official feature, I was too. But I have just built a custom endpoint extension that can both:

1) take an email and send it a magic link
2) receive a magic link token to validate and return a login token

Now I just need to validate that I can get it to work with my front end and properly set the cookie in the browser. 

My use case is with a custom front end, I would need an extension bundle with at least the form to input email for it  to work with the Directus App. But that's out of scope for me, at least now. 

I store the magic tokens in a dedicated table, to not pollute directus_users.

If there is any interest in this, I can try to get it to the Marketplace or just put it in a public GitHub repo.

**Reply by AndreasGBS** [reply]
HI @carwei, a marketplace version would be cool, thx in advance!

**Reply by TorbenHammes** [reply]
Hello, 
That sounds good. I would be happy about a marketplace plugin. I had also considered finding a solution via third party IDP (OAuth), but at this point I need to familiarise myself with the API first. The last time I worked with the API was with v9 ^^
But I don't think that was so well documented or doesn't really work with the SDK, as far as I remember... some new problems 😅.

Thank you in advance!

**Reply by TorbenHammes** [reply]
@carwei do you have any news for us?

**Reply by nicke** [reply]
@carwei I'm very interested on this. Do you have an update for us?

**Reply by carwei** [reply]
@TorbenHammes, @nicke and @AndreasGBS: I got it working pretty well and have used it internally for a while. I want to finish off some rough edges and documentation before I distribute it.

Are you guys fine with just the endpoint extension, i.e. using this with your own custom front-end or does it need to integrate with the Data Studio App (the official Directus UI)? Then I'll need to package it as a bundle to make room for an input form, so it's definitely an extra step and some UX considerations.

Edit: I realized my extension might not be "sandboxed" and that is a Marketplace requirement. I'll need to look into that!

**Reply by TorbenHammes** [reply]
The endpoint would be great for the beginning. Thanky for your work👌🏼

If you need some help you can share it with me. 

Greetings from Berlin✌🏼

**Reply by AndreasGBS** [reply]
@carwei The endpoint would be enough for me too, thank you very much!

**Reply by nicke** [reply]
@carwei endpoint is enough for me! Super!

**Reply by antontranelis** [reply]
I'm also looking forward to this extension! Thank you @carwei

**Reply by mattbisme** [reply]
Would this work via the Directus JavaScript SDK as well? I'm guessing not, but it would be great to integrate into an existing project.

**Reply by carwei** [reply]
> Would this work via the Directus JavaScript SDK as well? I'm guessing not, but it would be great to integrate into an existing project. 

@mattbisme, not sure what you mean with integration. It's an endpoint extension, so it's stand alone.

**Reply by antontranelis** [reply]
How is it going @carwei ? Do you need some help or feedback?

**Reply by carwei** [reply]
> How is it going @carwei ? Do you need some help or feedback?

@antontranelis I'm still on vacation, so I'm mostly AFK. I'll be back at the office soon though, and I'll spin up a public repo for this extension in the next 1-2 weeks. Are you able to run Directus self-hosted/locally? Then it would be great if you could give it a test run (before I figure out how to publish it to the Directus marketplace).

I appreciate all the confirmed interest, looking forward to sharing this.

**Reply by antontranelis** [reply]
Thanks for the update! I'm looking forward to test it as soon as it's ready

**Reply by nicke** [reply]
Hello, just a quick follow-up, I hope you don't mind. Is there any news yet on a sample extension? Thank you.

**Reply by carwei** [reply]
> Hello, just a quick follow-up, I hope you don't mind. Is there any news yet on a sample extension? Thank you.

Hi @nicke, I don't mind. Thanks for the reminder. It's been a little intense getting back to work, but I've now created a private repository for my extension. I want to figure out how to convert it to the "bundle" type (to support future iterations better) and double-check the documentation before I make it public. Hopefully, I'll be done by the end of next week (*fingers crossed*). When it's public, I'll really appreciate all your feedback before the next step: distributing it via the Marketplace.

**Reply by nicke** [reply]
@carwei,  I can give you a tip. Just use the Directus Extension Creator tool.

`npx create-directus-extension@latest` and choose "bundle".

Brian shows it here:
https://www.youtube.com/watch?v=zH3FEtebc7k&t=1695s

**Reply by carwei** [reply]
> @carwei,  I can give you a tip. Just use the Directus Extension Creator tool.
> 
> 
> 
> `npx create-directus-extension@latest` and choose "bundle".
> 
> 
> 
> Brian shows it here:
> 
> https://www.youtube.com/watch?v=zH3FEtebc7k&t=1695s

Great, I'll check it out. I've used that tool for a handful of extensions, just never created a bundle before.

**Reply by carwei** [reply]
@TorbenHammes @AndreasGBS @nicke @antontranelis Just a minor update. The code is in a private repository, I just want to give it a last test run before I invite you in. I updated the docs a bit, but you're more than welcome to validate them when you try it. I should say that it currently only works for SMTP, but I'm happy to extend that if needed. Hopefully, I can do the test run tonight or tomorrow.

**Reply by carwei** [reply]
@TorbenHammes @AndreasGBS @nicke @antontranelis I thought I could just do a test, but I realized I needed to give you more context. I thought of a way to provide a demo with both "request link" and "verify link" code, so I spent a considerable amount of time on that as a separate endpoint in the bundle. I hope it will be helpful to you, but let me know. Please give me feedback  (including documentation). It's my first open-source contribution/project 😄. Hopefully, I can bring it to the Marketplace if it works well. Sadly, making it work with "Data Studio" (instead of a custom front end) is very difficult, I've documented why in the README.md 

https://github.com/carwei/directus-magic-link-auth

**Reply by carwei** [reply]
@rijkvanzanten, feel free to give it a spin. Perhaps it answers some outstanding spec questions that were mentioned.

I don't see how I would be able to do Data Studio logins (only custom front ends), since I'm blocked by security measures to correctly set the required cookies for that.

**Reply by carwei** [reply]
@TorbenHammes, let me know if you got it working. I hope it makes sense 😄
